digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/service/lane/LaneService.scala

```diff
@@ -0,0 +1,609 @@
+
+    val existingAssets =
+      withDynTransaction {
+        dao.fetchLanesByLinkIdAndSideCode(linkId, sideCode)
+      }
+
+    existingAssets.map { lane =>
+    val geometry = GeometryUtils.truncateGeometry3D(roadLink.geometry, lane.startMeasure, lane.endMeasure)
+    val endPoints = GeometryUtils.geometryEndpoints(geometry)
+
+    PieceWiseLane(lane.id, lane.linkId, lane.sideCode, lane.expired, geometry,
+      lane.startMeasure, lane.endMeasure,
+      Set(endPoints._1, endPoints._2), lane.modifiedBy, lane.modifiedDateTime,
+      lane.createdBy, lane.createdDateTime,  roadLink.vvhTimeStamp,
+      lane.geomModifiedDate, roadLink.administrativeClass, lane.attributes, attributes = Map("municipality" -> lane.municipalityCode))
+    }
+  }
+
+
+  def persistProjectedLinearAssets(newLanes: Seq[PersistedLane]): Unit ={
+    if (newLanes.nonEmpty)
+      logger.info("Saving projected lanes")
+
+    val username = "projectedLanes"
+    val (toInsert, toUpdate) = newLanes.partition(_.id == 0L)
+
+    withDynTransaction {
+      if(toUpdate.nonEmpty) {
+
+        updatePersistedLanes(toUpdate, username)
+
+        if (newLanes.nonEmpty)
+          logger.info("Updated ids/linkids " + toUpdate.map(a => (a.id, a.linkId)))
+      }
+
+      toInsert.foreach{ lane =>
+         createWithoutTransaction( lane , username)
+      }
+
+      if (toInsert.nonEmpty)
+        logger.info("Added lanes for linkids " + newLanes.map(_.linkId))
+    }
+  }
+
+  def generateLanes (toCreate: Seq[PersistedLane]): Unit = {
+    if (toCreate.nonEmpty)
+      logger.info("Saving new lanes")
+
+    withDynTransaction {
+      toCreate.foreach{ lane =>
+        createWithoutTransaction( lane , "lanesGenerator")
+      }
+
+      if (toCreate.nonEmpty)
+        logger.info("Added lanes for linkids " + toCreate.map(_.linkId))
+    }
+  }
+
+
+  /**
+    * Returns lanes ids. Used by Digiroad2Api /lane POST and /lane DELETE endpoints.
+    */
+  def getPersistedLanesByIds( ids: Set[Long], newTransaction: Boolean = true): Seq[PersistedLane] = {
+    if(newTransaction)
+      withDynTransaction {
+        dao.fetchLanesByIds( ids )
+      }
+    else
+      dao.fetchLanesByIds( ids )
+  }
+
+  def getPropertyValue(newLane: NewIncomeLane, publicId: String) = {
+    val laneProperty = newLane.attributes.properties.find(_.publicId == publicId)
+                                            .getOrElse(throw new IllegalArgumentException(s"Attribute '$publicId' not found!"))
+
+    val finalValue = if (laneProperty.values.nonEmpty)
+                        laneProperty.values.head.value
+                      else
+                        None
+
+    finalValue
+  }
+
+  def getLaneCode (newIncomeLane: NewIncomeLane): String = {
+    val laneCodeValue = getPropertyValue(newIncomeLane, "lane_code")
+
+    val laneCode = if (laneCodeValue.toString.trim.nonEmpty)
+      laneCodeValue.toString.trim
+    else
+      throw new IllegalArgumentException("Lane code attribute not found!")
```

Comment:
None.toString.trim.nonEmpty can be true. If so this can be wrong, confirm this please

Reply:
Indeed.
It is better to validate that too

Reply:
Done

Reply:
if (laneCodeValue.isDefined && laneCodeValue.toString.trim.nonEmpty) should look better